### PR TITLE
PXS-597 Implement repository service pattern for experiences

### PR DESCRIPTION
**Pull request by qtipee**
The `Pager` object of the `ExperienceViewModel` now uses an `ExperienceService` and an `ExperienceRepository` ; it will simplify the test (Firestore mock) and we will also be able to easily change to another repository if needed in the future.

### CODE REVIEW

File: app/src/main/java/com/pindex/main/data/ExperiencePagingSource.kt [hunk 1]
```diff
@@ -1,40 +1,26 @@
...
+    override suspend fun load(params: LoadParams<List<ExperienceDto>>): LoadResult<List<ExperienceDto>, ExperienceDto> {
         return try {
-            // Current chunk of experiences to fetch
-            val currentPage = params.key ?: db.collection(Constants.FIRESTORE_EXPERIENCES_COLLECTION)
-                    .whereEqualTo("status", "listed")
-                    .limit(Constants.FIRESTORE_QUERY_LIMIT)
-                    .get()
-                    .await()
+            // If no key is passed, get the first chunk of data
+            val currentPage = params.key ?: service.getPage(Constants.FIRESTORE_QUERY_LIMIT)
```
Comment (xavierpantet) [inline app/src/main/java/com/pindex/main/data/ExperiencePagingSource.kt:17]: Rename `Constants.FIRESTORE_QUERY_LIMIT` to something more generic (e.g., `Constants.EXPERIENCES_LIST_PAGE_SIZE`).

Ideally, there should be no reference to Firestore in the code, except in the `FirebaseExperienceRepository`

File: app/src/main/java/com/pindex/main/data/ExperiencePagingSource.kt [hunk 2]
```diff
@@ -1,40 +1,26 @@
...
-
-            // Next chunk of experiences to fetch
-            val nextPage = db.collection(Constants.FIRESTORE_EXPERIENCES_COLLECTION)
-                    .whereEqualTo("status", "listed")
-                    .limit(Constants.FIRESTORE_QUERY_LIMIT)
-                    .startAfter(lastDocumentSnapshot)
-                    .get()
-                    .await()
+            // Next chunk of data
+            val nextPage = service.getPage(Constants.FIRESTORE_QUERY_LIMIT)
```
Comment (xavierpantet) [inline app/src/main/java/com/pindex/main/data/ExperiencePagingSource.kt:20]: Same comment here.

File: app/src/main/java/com/pindex/main/repositories/FirebaseExperienceRepository.kt [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    // Firestore instance
+    private val firestore: FirebaseFirestore = FirebaseFirestore.getInstance()
+
+    // Last experience used to fetch the next chunk of data
+    private var lastDocumentSnapshot: DocumentSnapshot? = null
+
+    override suspend fun getPage(limit: Long): List<ExperienceDto> {
+        var query = firestore.collection(Constants.FIRESTORE_EXPERIENCES_COLLECTION)
+                .whereEqualTo("status", "listed")
```
Comment (xavierpantet) [inline app/src/main/java/com/pindex/main/repositories/FirebaseExperienceRepository.kt:22]: Magic strings here. Move as class constants instead.

**Review by xavierpantet — CHANGES_REQUESTED** [review]

**Comment by xavierpantet** [comment]
I suggest we also move Firestore-related constants (`FIRESTORE_EXPERIENCES_COLLECTION`, `FIRESTORE_PAGE_SIZE`, `FIRESTORE_QUERY_LIMIT`) out of `Constants.tk` to move them into `FirebaseExperienceRepository`. The code should not be aware of Firebase, in general.

Also, do we need `FIRESTORE_PAGE_SIZE` or is it the same as `FIRESTORE_QUERY_LIMIT`?

File: app/src/main/java/com/pindex/main/data/ExperiencePagingSource.kt (hunk 1 above)
Comment (qtipee) [reply app/src/main/java/com/pindex/main/data/ExperiencePagingSource.kt:17]: Well this class is not used anymore as we are now using an `ExperienceRepository` --> gonna delete it!

File: app/src/main/java/com/pindex/main/data/ExperiencePagingSource.kt (hunk 2 above)
Comment (qtipee) [reply app/src/main/java/com/pindex/main/data/ExperiencePagingSource.kt:20]: Class deleted

**Comment by qtipee** [comment]
> I suggest we also move Firestore-related constants (`FIRESTORE_EXPERIENCES_COLLECTION`, `FIRESTORE_PAGE_SIZE`, `FIRESTORE_QUERY_LIMIT`) out of `Constants.tk` to move them into `FirebaseExperienceRepository`. The code should not be aware of Firebase, in general.
> 
> Also, do we need `FIRESTORE_PAGE_SIZE` or is it the same as `FIRESTORE_QUERY_LIMIT`?

- Firestore constants moved
- Only one constant for the number of experiences to fetch per request ; note that we have to convert it to an Int in the ViewModel

**Review by xavierpantet — APPROVED** [review]

**Pull request merged**